lib/carbon/writer.py

```diff
@@ -42,121 +54,174 @@
+        # reason.
+        continue
+
+      archiveConfig = None
+      xFilesFactor, aggregationMethod = None, None
+
+      for schema in SCHEMAS:
+        if schema.matches(metric):
+          if settings.LOG_CREATES:
+            log.creates('new metric %s matched schema %s' % (metric, schema.name))
+          archiveConfig = [archive.getTuple() for archive in schema.archives]
+          break
+
+      for schema in AGGREGATION_SCHEMAS:
+        if schema.matches(metric):
+          if settings.LOG_CREATES:
+            log.creates('new metric %s matched aggregation schema %s'
+                        % (metric, schema.name))
+          xFilesFactor, aggregationMethod = schema.archives
+          break
+
+      if not archiveConfig:
+        raise Exception(("No storage schema matched the metric '%s',"
+                         " check your storage-schemas.conf file.") % metric)
+
+      if settings.LOG_CREATES:
+        log.creates("creating database metric %s (archive=%s xff=%s agg=%s)" %
+                    (metric, archiveConfig, xFilesFactor, aggregationMethod))
       try:
-        t1 = time.time()
-        # If we have duplicated points, always pick the last. update_many()
-        # has no guaranted behavior for that, and in fact the current implementation
-        # will keep the first point in the list.
-        datapoints = dict(datapoints).items()
-        state.database.write(metric, datapoints)
-        updateTime = time.time() - t1
+        state.database.create(metric, archiveConfig, xFilesFactor, aggregationMethod)
+        state.database.tag(metric, httpRequest=httpRequest)
+        instrumentation.increment('creates')
       except Exception as e:
         log.err()
-        log.msg("Error writing to %s: %s" % (metric, e))
+        log.msg("Error creating %s: %s" % (metric, e))
         instrumentation.increment('errors')
-      else:
-        pointCount = len(datapoints)
-        instrumentation.increment('committedPoints', pointCount)
-        instrumentation.append('updateTimes', updateTime)
-        if settings.LOG_UPDATES:
-          log.updates("wrote %d datapoints for %s in %.5f seconds" % (pointCount, metric, updateTime))
+        continue
+
+    # If we've got a rate limit configured lets makes sure we enforce it
+    waitTime = 0
+    if UPDATE_BUCKET:
+      t1 = time.time()
+      yield UPDATE_BUCKET.drain(1, blocking=True)
+      waitTime = time.time() - t1
 
-    # Avoid churning CPU when only new metrics are in the cache
-    if not dataWritten:
-      time.sleep(1)
+    try:
+      t1 = time.time()
+      # If we have duplicated points, always pick the last. update_many()
+      # has no guaranted behavior for that, and in fact the current implementation
+      # will keep the first point in the list.
+      datapoints = dict(datapoints).items()
+      state.database.write(metric, datapoints)
+      if random.randint(1, settings.TAG_UPDATE_INTERVAL) == 1:  # nosec
```

Comment:
Isn't that a big usability issue ? It means that if you only send a few points to a metric it might not be created ? Or maybe I fail to understand why we need this here (add a comment in the code ?)

Reply:
There are 2 branches here, in the "new metric" branch `state.database.tag` is called unconditionally.  This branch is handling new data points for an existing series, it calls `state.database.tag` roughly every `settings.TAG_UPDATE_INTERVAL` points just to make sure that after upgrade, switching tag databases, or if there is an issue with the tagdb then the system will self-heal.

It would also be possible in the future for the tagdb to keep track of approximate first/last seen timestamps for each series so we can pass the time range through to the tagdb when searching via `seriesByTag`, but I have left that for later to keep things simple in this initial version.

Reply:
OK